pkg/cmd/dcos_autocomplete.go

```diff
@@ -0,0 +1,141 @@
+package cmd
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/dcos/dcos-cli/pkg/cli"
+	"github.com/spf13/cobra"
+	flag "github.com/spf13/pflag"
+)
+
+func newAutocompleteCommand(ctx *cli.Context) *cobra.Command {
+	cmd := &cobra.Command{
+		Use: "__autocomplete__",
+
+		// Hide it from the Usage output
+		Hidden: true,
+
+		// This needs to be true to prevent __autocomplete__ from trying to parse flags that are meant for
+		// the commands it's trying to complete causing it to fail.
+		DisableFlagParsing: true,
+		Args:               cobra.MinimumNArgs(1),
+		ValidArgs:          []string{"bash", "zsh"},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if len(args) == 0 {
+				return errors.New("no arguments given")
+			}
+			shell := args[0]
+
+			// root.Find will error out if it's given a flag that isn't accepted by the command it finds so
+			// we need to strip out the flags currently in the list to give just the command names to
+			// Find.
+			commands, flags := stripFlags(args[1:])
+			flagComplete := isFlag(args[len(args)-1])
+			root := cmd.Parent()
+
+			var completions []string
+
+			switch shell {
+			case "zsh":
+			case "bash":
+				command, _, err := root.Find(commands)
+				if err != nil {
+					// if no command was found, default to the root
+					command = root
+				}
+
+				// We'll need to determine whether or not the found command is internal or external here.
+				// To do that we use the Annotations object which the cobra documentation doesn't say much
+				// about but appears to be for arbitrary metadata so we can use that to hold whether or
+				// not a command is internal or external. In this case we're only interested in whether the
+				// external key exists because we assume the default is internal.
+				if _, exists := cmd.Annotations["external"]; exists {
+					completions = externalCompletion(command)
+				} else {
+					completions = internalCompletion(command, flags, flagComplete)
+				}
+
+				completions = append(completions, customCompletion(command, ctx)...)
+			default:
+				return fmt.Errorf("no completion implemented for shell %s", shell)
+			}
+
+			for _, c := range completions {
+				fmt.Fprintln(ctx.Out(), c)
+			}
+
+			return nil
+		},
+	}
+
+	return cmd
+}
+
+func isFlag(arg string) bool {
+	return strings.HasPrefix(arg, "-")
```

Comment:
I hadn't thought about that. It definitely doesn't. I'm not even sure if it would be reasonably possible to handle that case actually. To know if a flag takes an argument I would need to know what command that flag is tied to but we can't ignore this because it would add admin in as part of the commands.

Not sure how to solve this one yet.

Reply:
IIUC this method is used to strip flags basically, could you elaborate a bit more on why it's needed to strip flags. I've seen `root.Find will error out if it's given a flag that isn't accepted by the command` but not sure I fully get this, do you mean it fails when it's being passed plugin subcommand flags (as it doesn't know about these) or when being passed wrong flags to internal subcommands?

Reply:
Looks to be working fine without that. I think I mixed it up with it failing if autocomplete didn't have `DisableFlagParsing`. I'll take that out. The example of `--flag value` though shows that how this handles flag completion is incomplete.